src/tests/integration/SparkLend.t.sol

```diff
@@ -0,0 +1,400 @@
+        
+        // Give us some DAI
+        dai.setBalance(address(this), buffer * 100000000);
+
+        // Deposit WETH into the pool
+        uint256 amt = 1_000_000 * WAD;
+        DSTokenAbstract weth = DSTokenAbstract(dss.getIlk("ETH", "A").gem);
+        weth.setBalance(address(this), amt);
+        weth.approve(address(sparkPool), type(uint256).max);
+        dai.approve(address(sparkPool), type(uint256).max);
+        sparkPool.supply(address(weth), amt, address(this), 0);
+
+        assertGt(getDebtCeiling(), 0);
+
+        // Recompute the dai interest rate strategy to ensure the new line is taken into account
+        daiInterestRateStrategy.recompute();
+
+        basePostSetup();
+    }
+
+    // --- Overrides ---
+    function adjustDebt(int256 deltaAmount) internal override {
+        if (deltaAmount == 0) return;
+        
+        int256 newBuffer = int256(plan.buffer()) + deltaAmount;
+        vm.prank(admin); plan.file("buffer", newBuffer >= 0 ? uint256(newBuffer) : 0);
+        hub.exec(ilk);
+    }
+
+    function adjustLiquidity(int256 deltaAmount) internal override {
+        if (deltaAmount == 0) return;
+
+        if (deltaAmount > 0) {
+            // Supply to increase liquidity
+            uint256 amt = uint256(deltaAmount);
+            dai.setBalance(address(this), dai.balanceOf(address(this)) + amt);
+            sparkPool.supply(address(dai), amt, address(0), 0);
+        } else {
+            // Borrow to decrease liquidity
+            uint256 amt = uint256(-deltaAmount);
+            if (amt > getLiquidity()) {
+                amt = getLiquidity();
+            }
+            sparkPool.borrow(address(dai), amt, 2, 0, address(this));
+        }
+    }
+
+    function generateInterest() internal override {
+        // Generate interest by borrowing and repaying
+        uint256 performanceBonus = daiInterestRateStrategy.performanceBonus();
+        sparkPool.supply(address(dai), performanceBonus * 4, address(this), 0);
+        sparkPool.borrow(address(dai), performanceBonus * 2, 2, 0, address(this));
+        vm.warp(block.timestamp + 1 days);
+        sparkPool.repay(address(dai), performanceBonus * 2, 2, address(this));
+        sparkPool.withdraw(address(dai), performanceBonus * 4, address(this));
+    }
+
+    function getLiquidity() internal override view returns (uint256) {
+        return dai.balanceOf(address(adai));
+    }
+
+    // --- Helper functions ---
+    function getDebtCeiling() internal view returns (uint256) {
+        (,,, uint256 line,) = dss.vat.ilks(ilk);
+        return line;
+    }
+
+    function getDebt() internal view returns (uint256) {
+        (, uint256 art) = dss.vat.urns(ilk, address(pool));
+        return art;
+    }
+
+    function _divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        unchecked {
+            z = x != 0 ? ((x - 1) / y) + 1 : 0;
+        }
+    }
+
+    function getSupplyUsed(address asset) internal view returns (uint256) {
```

Comment:
Isn't this calculation just the total supply? I mean I see it used here where I'd expect the total supply to be used - https://github.com/aave/aave-v3-core/blob/94e571f3a7465201881a59555314cd550ccfda57/contracts/protocol/libraries/logic/ValidationLogic.sol#L74